### ISSUE TITLE
High precision and gradient discrepancy in RWKV Triton implementation between chunk and recurrent_fuse

### ISSUE THREAD

**Issue by uniartisan** [sustcsonglin/flash-linear-attention#36]
![image](https://github.com/user-attachments/assets/ccd54a2b-bb0e-4833-a925-9a46e614df4f)

I've noticed a significant discrepancy in precision and gradients in the chunk implementation of the RWKV operator in the Triton implementation. The file in question is `flash-linear-attention/fla/ops/rwkv6/chunk_naive.py`.

Important observation:
The standard Triton implementation and the PyTorch implementation show no significant precision differences. However, the chunk implementation exhibits extremely high discrepancies.

Details:
- About 90% of the outputs show large differences in the chunk implementation.
- about 10%-200% of the gradient differences in the chunk implementation.
- The maximum differences observed are as follows:

```python
tensor(145., device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>)
tensor(145., device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>)
tensor(202., device='cuda:0', dtype=torch.bfloat16)
tensor(198., device='cuda:0', dtype=torch.bfloat16)
tensor(57.2500, device='cuda:0', dtype=torch.bfloat16)
tensor(163., device='cuda:0', dtype=torch.bfloat16)
tensor(1392., device='cuda:0', dtype=torch.bfloat16)
```

These discrepancies are concerning as they could lead to significant differences in model performance and training stability, especially when using the chunk implementation.

Questions:
1. Is this level of difference expected in the chunk implementation, given that the standard Triton and PyTorch implementations align well?
2. Could there be a specific bug or issue in the chunk implementation causing these large discrepancies?
3. Are there any known numerical stability issues in the chunking process that could explain these differences?
4. How does the chunking process differ from the standard implementation, and could this difference account for the observed discrepancies?

Steps to reproduce:
```
python flash-linear-attention/fla/ops/rwkv6/chunk_naive.py
```
Environment:
- PyTorch version: 2.3
- GPU model: NVIDIA 4090

- Pytorch 2.4
- GPU intel arc a770

Any insights or suggestions for investigating this issue further would be greatly appreciated. In particular, any guidance on debugging the chunk implementation or understanding why it differs so significantly from the other implementations would be very helpful. Thank you for your time and attention to this matter.

**Comment by uniartisan** [comment]
solution: 
```
o2, _ = chunk_rwkv6(q, k, v, w.clone(), u, scale=1.0)
```

**Comment by uniartisan** [comment]
```
if __name__ == "__main__":
    from fla.utils import get_available_device
    device = get_available_device()
    B = 4
    H = 4
    L = 4096
    D = 100
    dtype = torch.bfloat16
    require_grad = True
    q = (torch.randn(B, H, L, D).to(device).to(dtype)).requires_grad_(require_grad)
    k = (torch.randn(B, H, L, D).to(device).to(dtype)).requires_grad_(require_grad)
    v = torch.randn(B, H, L, 2*D).to(device).to(dtype).requires_grad_(require_grad)
    w = torch.nn.functional.logsigmoid(torch.randn(B, H, L, D)).to(device).to(dtype).requires_grad_(require_grad)
    u = (torch.randn(H, D).to(device).to(dtype)).requires_grad_(require_grad)
    h = torch.randn(B, H, D, 2*D, device=device, dtype=dtype, requires_grad=True)
    do = torch.rand_like(v).to(device)
    o2, _ = chunk_rwkv6(q, k, v, w.clone(), u,initial_state=h, scale=1.0)
    o, _ = fused_recurrent_rwkv6(q, k, v, w, u,initial_state=h, scale=1.0)

    o.backward(do) #正确的算法
    dq, q.grad = q.grad.clone(), None
    dk, k.grad = k.grad.clone(), None
    dv, v.grad = v.grad.clone(), None
    dw, w.grad = w.grad.clone(), None
    du, u.grad = u.grad.clone(), None
    dh, h.grad = h.grad.clone(), None
    # print((o - o2).abs().mean())
    o2.backward(do) # 错误的算法
    # print((o-o2).abs().max())
    # print((q.grad - dq).abs().max())
    # print((k.grad - dk).abs().max())
    # print((v.grad - dv).abs().max())
    # print((w.grad - dw).abs().max())
    # print((u.grad - du).abs().max())
    # print((o-o2).abs()/o.abs())
    print((h.grad-dh).abs().max())
    print(((h.grad-dh).abs()/dh.abs()).max())
```
tensor(0.0312, device='xpu:0', dtype=torch.bfloat16)
tensor(3024., device='xpu:0', dtype=torch.bfloat16)
